Isaacs/lifecycle envs

isaacs opened the pull request:
Use `@npmcli/run-script` for the run-script command, and add a few missing configs to flatOptions.

isaacs committed:
Remove HTTP Referer header config

Last bit of https://github.com/npm/cli/issues/930

isaacs committed:
Set basic npm envs in main command function

This sets environment variables for npm_command, npm_execpath,
npm_node_execpath, NODE_OPTIONS, and any non-default configs that do not
start with @, /, or _ (as those are potentially sensitive, or at least
not useful).

This is in preparation for the move from npm-lifecycle to
@npmcli/run-script for running lifecycle scripts.

isaacs committed:
@npmcli/run-script@1.3.0

isaacs committed:
nopt@4.0.3

isaacs committed:
Set non-default configs in the environment

Setting up for the switch from npm-lifecycle to @npmcli/run-script.

isaacs committed:
Use @npmcli/run-script for npm run

isaacs committed:
flatOptions: include preferDedupe and ignoreScripts

isaacs committed:
fix tests for flatOptions with recent changes

mikemimik commented on lib/run-script.js:
```diff
@@ -75,111 +116,51 @@ function list (cb) {
-      }
-    }
-    cmds = [cmd]
+  const allScripts = scripts ? Object.keys(scripts) : {}
```
If `scripts` is _truthy_ then you get an array of strings, if `scripts` is _falsy_ you get an empty object. Seems like this should be the same value.

On line `146` we are using a `for...of` statement which is meant for array's and not objects (it will throw if it gets an object)

isaacs replied:
Confirmed, this is an oversight.

mikemimik commented on lib/run-script.js:
```diff
@@ -75,111 +116,51 @@ function list (cb) {
+  const prefix = '  '
+  const cmds = []
+  const runScripts = []
+  for (const script of allScripts) {
```
Could be a problem (from comment above)

mikemimik commented on lib/run-script.js:
```diff
@@ -75,111 +116,51 @@ function list (cb) {
+  const cmds = []
+  const runScripts = []
+  for (const script of allScripts) {
+    const list = cmdList.includes(script) ? cmds : runScripts
```
Both `cmds` and `runScripts` start as an empty array. I'm not sure why we need to check if `cmdList` has the `script` in it, if in both cases we return an empty array.

mikemimik commented on lib/run-script.js:
```diff
@@ -75,111 +116,51 @@ function list (cb) {
+  for (const script of runScripts) {
+    output(prefix + script + indent + scripts[script])
+  }
+  return allScripts
```
The conditional `if` logic above is a little confusing. I now realise the point of the `cmdList.includes()` is to find out if what you typed as a user was, for example, `npm run test`. That output should be different than if you ran, for example, `npm run myscript`.

isaacs replied:
This is for listing out scripts that are runnable as a command differently from those that _have_ to be run with `npm run`.

Eg:

```
$ node . run
Lifecycle scripts included in npm:
  preversion
    bash scripts/update-authors.sh && git add AUTHORS && git commit -m "update AUTHORS" || true
  test
    npm run test-tap --
  posttest
    npm run lint

available via `npm run-script`:
  dumpconf
    env | grep npm | sort | uniq
  prepare
    node bin/npm-cli.js rebuild && node bin/npm-cli.js --no-audit --no-timing prune --prefix=. --no-global && rimraf test/*/*/node_modules && make -j4 mandocs
  ...
```

mikemimik replied:
I think I understand. I guess the distinction wasn't clear when reading through the code. I'm not sure if that's indicative of me needing more context or the code not being clear enough.

isaacs replied:
When you run `npm run` without any args, it calls the `list` function, which dumps a list of which scripts are available.

The conditional on 147 is deciding which list to push the command name into, and then we loop over both lists to print out which ones are command lifecycles, and which have to be run explicitly.

mikemimik commented on lib/config/flat-options.js:
```diff
@@ -159,7 +163,7 @@ const flatOptions = npm => npm.flatOptions || Object.freeze({
   globalStyle: npm.config.get('global-style'),
   legacyBundling: npm.config.get('legacy-bundling'),
-  scriptShell: npm.config.get('script-shell'),
+  scriptShell: npm.config.get('script-shell') || undefined,
```
Question: What's the default returned value of `npm.config.get('...')` which doesn't find the key `...`?

isaacs replied:
Can be literally anything.

```
$ npm config get script-shell --script-shell=420lol
420lol
```

But @npmcli/run-script only allows Boolean or undefined.

isaacs replied:
Er, only allows strings.  (Sorry, was thinkign of another flag.)

So this is the case that would be a problem:

```
npm run foo --no-script-shell
```

And the effect of setting that `false` value to `undefined` is to fall back to the system default (`/bin/sh` or `cmd.exe`).

mikemimik commented on lib/config/flat-options.js:
```diff
@@ -57,7 +60,8 @@ const flatOptions = npm => npm.flatOptions || Object.freeze({
 
-  metricsRegistry: npm.config.get('metrics-registry'),
+  metricsRegistry: npm.config.get('metrics-registry') ||
+    npm.config.get('registry'),
```
Question: Are these the same thing?

isaacs replied:
Unless specified explicitly, `metrics-registry` defaults to the public registry.

mikemimik commented on lib/config/set-envs.js:
```diff
@@ -0,0 +1,58 @@
+//
+// See https://github.com/npm/rfcs/pull/90
+
+const envName = name => `npm_config_${name.replace(/-/g, '_')}`
```
nit: would be nice if we had a rule of thumb that any time you use regexp, no matter how simple, you have to comment what you're intent is with it.

isaacs replied:
In more elaborate cases, I might agree, but as a rule, that seems to just tend towards excessive commenting.

```js
// replaces - with _
str.replace(/-/g, '_')
```

which is in the same category as:

```js
// adds 1 to foo
foo += 1
```

Why would we need this for trivial regexps and not for every other trivial thing?

mikemimik replied:
I agree with this.

mikemimik commented on lib/config/set-envs.js:
```diff
@@ -0,0 +1,58 @@
+const envName = name => `npm_config_${name.replace(/-/g, '_')}`
+
+const envKey = (key, val) => {
+  return !/^[\/@_]/.test(key) &&
```
nit: regexp

mikemimik replied:
Given the above, does this just mean: String starting with `/@_`?

isaacs replied:
We're filtering out all strings that start with slash, at, or underscore, yes.  Because those are potentially sensitive.

mikemimik commented on lib/config/set-envs.js:
```diff
@@ -0,0 +1,58 @@
+const envKey = (key, val) => {
+  return !/^[\/@_]/.test(key) &&
+    (typeof val === 'string' || Array.isArray(val)) &&
+    `npm_config_${key.replace(/-/g, '_').toLowerCase()}`
```
nit: regexp

mikemimik commented on lib/npm.js:
```diff
@@ -113,19 +113,20 @@
             'See the README.md or bin/npm-cli.js for example usage.'
```
Line 106: `var registryRefer` is defined by never used
Line 95: `var littleGuys` is a reference to incorrect spelling, but `lib/config/cmd-list.js` already has an `affordances` section; shouldn't we be using that?

isaacs replied:
Yes, registryRefer should be removed, since we're not sending the referer header any more.

Istr there's a difference between "affordances" and "littleGuys", but I don't recall it offhand.  Will dig in and sort that out.

isaacs replied:
Nope, no difference.  Just apparently something that got overlooked 4 years ago when it stopped being relevant.

mikemimik reviewed (commented):
Couple things I found:
- small nits about using regexp
- few questions around config we're setting in `lib/config/flat-options.js`
- out of file scope cleanup in `lib/npm.js`
- logical questions in `lib/run-script.js`

isaacs committed:
Remove littleGuys, no longer relevant

The 'littleGuys' (ie, isntall, verison, udpate) are included in the
cmdList.affordances section, so are no longer printed in the default
help banner anyway.  We can remove this variable.

isaacs committed:
Fix (with hacky temporary code in some cases) usage output for all commands

isaacs commented:
Landed as part of #1021
